### PR TITLE
Strain

### PR DESCRIPTION
**Pull request by samwaseda**
Add Lagrangian strain

### CODE REVIEW

File: pyiron_atomistics/atomistics/structure/atoms.py [hunk 1]
```diff
@@ -2274,29 +2274,45 @@ def set_constraint(self, constraint=None):
...
             structure_copy = self.copy()
         else:
             structure_copy = self
         cell = structure_copy.cell.copy()
-        cell = np.matmul(epsilon + np.eye(3), cell)
+        if mode == 'linear':
+            F = epsilon + np.eye(3)
+        else:
+            E, V = np.linalg.eigh(2*epsilon+np.eye(3))
+            F = np.einsum('ik,k,jk->ij', V, np.sqrt(E), V)
```
Comment (jan-janssen) [inline pyiron_atomistics/atomistics/structure/atoms.py]: Can we have all modifications in the same set of lines? then its hopefully easier to debug in future. 
```suggestion
        if mode == 'linear':
            F = epsilon + np.eye(3)
        elif mode == 'lagrangian':
            if not np.allclose(epsilon, epsilon.T):
                raise ValueError("Strain must be symmetric if `mode = 'lagrangian'`")
            E, V = np.linalg.eigh(2*epsilon+np.eye(3))
            F = np.einsum('ik,k,jk->ij', V, np.sqrt(E), V)
        else:
            raise ValueError('mode must be `linear` or `lagrangian`')
```

Comment (samwaseda) [reply pyiron_atomistics/atomistics/structure/atoms.py]: ah ok true

File: pyiron_atomistics/atomistics/structure/atoms.py [hunk 2]
```diff
@@ -2274,29 +2274,49 @@ def set_constraint(self, constraint=None):
...
+                multiplied by a unit matrix.
             return_box (bool): whether to return a box. If set to True, only the returned box will
-                               have the desired strain and the original box will stay unchanged.
+                have the desired strain and the original box will stay unchanged.
+            mode (str): `linear` or `lagrangian`. If `linear`, `F` is equal to the epsilon - 1.
+                If `lagrangian`, epsilon is given by `(F^T * F - 1) / 2`. It raises an error if
+                the strain is not symmetric (if the shear components are given).
         """
+        if mode not in ['lagrangian', 'linear']:
+            raise ValueError('mode must be `linear` or `lagrangian`')
```
Comment (jan-janssen) [inline pyiron_atomistics/atomistics/structure/atoms.py]:
```suggestion
```

File: pyiron_atomistics/atomistics/structure/atoms.py [hunk 3]
```diff
@@ -2274,29 +2274,49 @@ def set_constraint(self, constraint=None):
...
         epsilon = np.array([epsilon]).flatten()
         if len(epsilon) == 3 or len(epsilon) == 1:
             epsilon = epsilon*np.eye(3)
         epsilon = epsilon.reshape(3, 3)
         if epsilon.min() < -1.0:
             raise ValueError("Strain value too negative")
+        if mode == 'lagrangian' and not np.allclose(epsilon, epsilon.T):
+            raise ValueError("Strain must be symmetric if `mode = 'lagrangian'`")
```
Comment (jan-janssen) [inline pyiron_atomistics/atomistics/structure/atoms.py]:
```suggestion
```

File: pyiron_atomistics/atomistics/structure/atoms.py (hunk 2 above)
Comment (jan-janssen) [reply pyiron_atomistics/atomistics/structure/atoms.py]: @samwaseda This part can also be removed

**Review by jan-janssen — APPROVED** [review]
Looks good to me

**Pull request merged**